[SPARK-52780] Add ToLocalIterator and Arrow Record Streaming

caldempsey opened the pull request:
### What changes were proposed in this pull request?

This PR adds streaming support to the Spark Connect Go client by introducing `ToLocalIterator(ctx)` on the DataFrame API, matching the functionality available in Scala's implementation. The implementation includes a new `types.RowIterator` interface that streams Arrow records batch-by-batch, and a `ToRecordBatches()` method on `ExecutePlanClient` that returns channels for streaming Arrow records and errors. Context cancellation is handled throughout the streaming pipeline to ensure proper resource cleanup.

### Why are the changes needed?

The Go client currently only supports `Collect()`, which loads all rows into memory at once. This approach doesn't work for large OLAP queries that return gigabytes of data, and prevents lightweight services like API gateways or serverless functions from using Spark Connect effectively.

This PR brings the same streaming capability that Scala and Java clients have through `toLocalIterator()`. Rows are deserialised batch-by-batch and released immediately after consumption. Since `ToTable` was pulling entire tables of Arrow batches, we needed a new method to stream batches individually. For now, I've reused `ReadArrowTableToRows` for the batch-to-row conversion, though this could be optimised in a follow-up PR. 

The changes are purely additive to maintain backward compatibility. External consumers can pull data at their own pace, making it suitable for streaming over HTTP or gRPC. I believe there are many opportunities to improve performance by reducing allocs and GC pressure, which can also be addressed in a follow up.

### Does this PR introduce any user-facing change?

Yes, it adds a new optional API:

```go
it, err := df.ToLocalIterator(ctx)
if err != nil {
    return err
}
defer it.Close()

for {
    row, err := it.Next()
    if err == io.EOF { 
        break 
    }
    if err != nil {
        return err
    }
    // process row ...
}
```

Existing code using `Collect()` or `ToArrow()` remains unchanged.

### How was this patch tested?

Unit tests verify that streaming returns rows in the correct order, handles EOF properly, cancels contexts gracefully without goroutine leaks, converts Arrow batches correctly, and propagates errors through the streaming pipeline.

The implementation was also tested against a Unity Catalog/Databricks Spark cluster using this streaming HTTP handler:

```go
// Example usage with proper context handling in a REST server
func StreamingHandler(w http.ResponseWriter, r *http.Request) {
    // Create a context with timeout for the request
    ctx, cancel := context.WithTimeout(r.Context(), 30*time.Minute)
    defer cancel()

    // Set up proper cleanup on client disconnect
    notify := w.(http.CloseNotifier).CloseNotify()
    go func() {
        select {
        case <-notify:
            cancel() // Cancel context if client disconnects
        case <-ctx.Done():
            // Context already done
        }
    }()

    var df sql.DataFrame // Your DataFrame
    
    rowIter, err := df.ToLocalIterator(ctx)
    if err != nil {
        http.Error(w, err.Error(), http.StatusInternalServerError)
        return
    }
    defer func() {
        if closeErr := rowIter.Close(); closeErr != nil {
            log.Printf("Error closing iterator: %v", closeErr)
        }
    }()

    w.Header().Set("Content-Type", "application/x-ndjson")
    w.WriteHeader(http.StatusOK)

    encoder := json.NewEncoder(w)
    
    for {
        select {
        case <-ctx.Done():
            // Context cancelled - exit gracefully
            return
        default:
        }

        row, err := rowIter.Next()
        if err != nil {
            if errors.Is(err, io.EOF) {
                return // Normal completion
            }
            if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
                return // Context cancellation - exit gracefully
            }
            // Log error but don't crash
            log.Printf("Error reading row: %v", err)
            return
        }

        // Stream the row as JSON
        if err := encoder.Encode(map[string]interface{}{
            "values": row.Values(),
        }); err != nil {
            // Client disconnected or other write error
            return
        }

        // Flush the response to ensure streaming
        if flusher, ok := w.(http.Flusher); ok {
            flusher.Flush()
        }
    }
}
```

All tests pass locally with `go test ./...`.

caldempsey commented on spark/client/base/base.go:
```diff
@@ -48,5 +48,6 @@ type SparkConnectClient interface {
 
 type ExecuteResponseStream interface {
 	ToTable() (*types.StructType, arrow.Table, error)
+	ToRecordBatches(ctx context.Context) (<-chan arrow.Record, <-chan error, *types.StructType)
```
Added so we can consume Apache Arrow record batches rather than the entire table at once, in truth, ToTable should re-use private methods w/ ToRecordBatches, but I've avoided this to avoid breaking clients for now.

caldempsey replied:
Thanks! I'll apply this comment (and all the others) in the next round of commits 🥳

caldempsey replied:
Updated to use Seq2

caldempsey commented on spark/sql/types/arrow.go:
```diff
@@ -68,6 +68,12 @@ func ReadArrowTableToRows(table arrow.Table) ([]Row, error) {
 	return result, nil
 }
 
+func ReadArrowRecordToRows(record arrow.Record) ([]Row, error) {
+	table := array.NewTableFromRecords(record.Schema(), []arrow.Record{record})
+	defer table.Release()
+	return ReadArrowTableToRows(table)
+}
+
```
Seems a bit inefficient to create a whole array.NewTableFromRecords. We should be able to go straight from records to rows, but I've left this for convenience sake.

caldempsey reviewed (commented):
There are a number of places we can improve the performance of the Spark Connect Go Client where it matters, namely by pre-allocating resources and re-using pointers to those allocations to reduce GC pressure. I think this could speed things up _substantially_.

caldempsey commented:
Seem to be some issues, will resolve.

grundprinzip commented:
I will take a look at the PR. It would be great to compare to https://github.com/apache/spark-connect-go/pull/103 and see how the native go iterators are different from what you propose here. 

Thanks for your contributions!

caldempsey commented on spark/client/client.go:
```diff
@@ -434,6 +443,151 @@ func (c *ExecutePlanClient) ToTable() (*types.StructType, arrow.Table, error) {
+
+				// Try to send the record, but respect context cancellation
+				select {
+				case recordChan <- record:
+					// Successfully sent
```
All this effort just so we can do this. Again, we should opt in to a DRY implementation with ToTable.

caldempsey replied:
left this for now, I think we can punt this down the road, mainly because I don't think it needs to be the focus of this PR but maybe the focus of a following one

caldempsey commented:
@grundprinzip Hey! I worked out the panic. It looks like the error I mentioned before only occurs when running the Databricks ODBC driver adjacent to it. There has to be some kind of signal handling collision between Spark Connect and that instance. Removing the Databricks ODBC driver instance from my main thread (a pretty big testing app) resolved the  issue. 

```
signal 13 received but handler not on signal stack
mp.gsignal stack [0xc000584000 0xc00058c000], mp.g0 stack [0x7fffb07f9280 0x7fffb0ff8e80], sp=0xc000815e20
fatal error: non-Go code set up signal handler without SA_ONSTACK flag
```

The introduction of context cancellation to Spark Connect only came in at about v40, and I was running v34 before, which means I think it might be worth testing the combination of cancelled context in Spark Connect on v40 with ODBC running adjacent. A pretty rare edge case, I know, but you seem to work at Databricks and might want to shout at the guys at Simba to make sure they are handling [Signal](https://go.dev/pkg/os/signal/?m=old) gracefully. Tagging @alexott onto this one. 

I think we can unblock this PR right now. I'll see if I can get a core dump to prove out the issue being related to Simba Spark ODBC, but the problem cannot be reproduced without it, and the PR otherwise seems feature complete.

caldempsey reviewed (commented):
@grundprinzip Hey! I worked out the panic. It looks like the error I mentioned before only occurs when running the Databricks ODBC driver adjacent to it. There has to be some kind of signal handling collision between Spark Connect and that instance. Removing the Databricks ODBC driver instance from my main thread (a pretty big testing app) resolved the  issue. 

```
signal 13 received but handler not on signal stack
mp.gsignal stack [0xc000584000 0xc00058c000], mp.g0 stack [0x7fffb07f9280 0x7fffb0ff8e80], sp=0xc000815e20
fatal error: non-Go code set up signal handler without SA_ONSTACK flag
```

The introduction of context cancellation to Spark Connect only came in at about v40, and I was running v34 before, which means I think it might be worth internal testing the combination of cancelled context in Spark Connect on v40 with the Databricks ODBC Driver. A pretty rare edge case, I know, but you seem to work at Databricks and might want to shout at the guys at Simba to make sure they are handling [Signal](https://go.dev/pkg/os/signal/?m=old) gracefully. Tagging @alexott onto this one. 

I think we can unblock this PR right now. I'll see if I can get a core dump to prove out the issue being related to Simba Spark ODBC, but the problem cannot be reproduced without it, and the PR otherwise seems feature complete.

grundprinzip commented on spark/client/client.go:
```diff
@@ -368,6 +368,15 @@ func (c *ExecutePlanClient) ToTable() (*types.StructType, arrow.Table, error) {
 	c.done = false
 	for {
 		resp, err := c.responseStream.Recv()
+		if err != nil {
+			fmt.Printf("DEBUG: Recv error: %v, is EOF: %v\n", err, errors.Is(err, io.EOF))
+		}
+		if err == nil && resp != nil {
+			fmt.Printf("DEBUG: Received response type: %T\n", resp.ResponseType)
+			if _, ok := resp.ResponseType.(*proto.ExecutePlanResponse_ResultComplete_); ok {
+				fmt.Println("DEBUG: Got ResultComplete!")
+			}
+		}
```
Yeah, conceptually it would be better to add them to the appropriate client for example. In Spark 3.4 the stream was blocking and in 4.0 the stream by default is re-attacheable and the client has a specific implementation for that.

caldempsey replied:
i think we should leave this for now. good to resolve?

grundprinzip commented on spark/client/base/base.go:
```diff
@@ -48,5 +48,6 @@ type SparkConnectClient interface {
 
 type ExecuteResponseStream interface {
 	ToTable() (*types.StructType, arrow.Table, error)
+	ToRecordBatches(ctx context.Context) (<-chan arrow.Record, <-chan error, *types.StructType)
```
Would be good to add comments on the how this is intended to be used and what the difference to the `ToTable` is.

caldempsey replied:
100% I'll add this to the next commit.

caldempsey replied:
Added

grundprinzip commented on spark/client/client.go:
```diff
@@ -434,6 +443,151 @@ func (c *ExecutePlanClient) ToTable() (*types.StructType, arrow.Table, error) {
+	go func() {
+		defer func() {
+			// Ensure channels are always closed to prevent goroutine leaks
+			close(recordChan)
+			close(errorChan)
```
Does close require any error handling or is this final state.

caldempsey replied:
Final state here. Once closed, no more sends are allowed (sending on a closed channel panics), but you can still receive any buffered items, and subsequent receives yield the zero value immediately. Then attempting to close a closed channel will panic.

caldempsey replied:
this is outdated now, we use seq2 for pulls which simplifies a bit

grundprinzip commented:
@caldempsey, thanks a lot for your contribution! I will take another look in early next week, but I don't see a lot of contentious pieces.

caldempsey commented on spark/sql/dataframe.go:
```diff
@@ -213,6 +213,7 @@ type DataFrame interface {
 	Take(ctx context.Context, limit int32) ([]types.Row, error)
 	// ToArrow returns the Arrow representation of the DataFrame.
 	ToArrow(ctx context.Context) (*arrow.Table, error)
+	ToLocalIterator(ctx context.Context) (types.RowIterator, error)
```
Yeah, I'd prefer to rename this. I don't like the name `ToLocalIterator` because it doesn't tell us anything about the underlying mechanism. When I tested with Scala, I was surprised to find I can use ToLocalIterator in Scala Spark Connect on streams of Arrow batches. `IterArrowStream` or `StreamRows` would be so much better. Let me know what you prefer, I'm glad we're on the same page here (for different reasons)!

zeroshade replied:
I would be in favor of using `iter.Seq` and `iter.Seq2` 

For reference, Arrow also allows using `iter` directly: https://pkg.go.dev/github.com/apache/arrow-go/v18/arrow/array#IterFromReader and https://pkg.go.dev/github.com/apache/arrow-go/v18/arrow/array#ReaderFromIter

caldempsey replied:
Updated to use Seq2, seemed to better match the Iterator semantics we wanted to expose

caldempsey commented:
I looked into the panic `signal 13 received but handler not on signal stack... fatal error: non-Go code set up signal handler without SA_ONSTACK flag` that was happening in my application w/ ODBC and Spark Connect loaded side by side. Turns out it only occurs when running the Databricks ODBC driver alongside Spark Connect in the same process - run either one alone and everything works fine.

I added some socket monitoring to see what was going on and found both libraries maintain their own connections to Databricks through port 443:

```
ODBC Connections:    172.17.0.2:42562->44.234.192.45:443
                     172.17.0.2:42568->44.234.192.45:443
Spark Connect:       172.17.0.2:60452->44.234.192.46:443
                     172.17.0.2:41742->44.234.192.45:443
```

I _**THINK**_ the issue is becaauuuuse POSIX signals get delivered to the entire process, not individual sockets. I'm guessing the ODBC driver (`/opt/simba/spark/lib/64/libsparkodbc_sb64.so`) installs signal handlers when it loads, but doesn't use the `SA_ONSTACK` flag that Go requires. So when either library hits a broken pipe, the kernel sends SIGPIPE to the process, the ODBC handler catches it, and Go panics because the handler isn't set up correctly for Go's stack management.

With Spark Connect v40's new streaming and context cancellation support, SIGPIPE happens more often - cancel a context mid-stream and you'll likely trigger it. The ODBC driver's process-wide signal handler intercepts these signals regardless of which library's socket actually broke, while Spark Connect seems to handle the termination gracefully. The 'fix' here is really sweeping the issue under the rug by adding `signal.Ignore(syscall.SIGPIPE)` at startup e.g.

```
package main 

func init() {
	signal.Ignore(syscall.SIGPIPE)
}
```

When you call signal.Ignore(syscall.SIGPIPE), the entire process ignores SIGPIPE signals. The kernel simply doesn't deliver them to the Go process at all. Both libraries already handle EPIPE errors properly through return values, so they don't need the signal. A C++ dev I know (who I can't mention) said this is actually the modern way to handle broken pipes - check return values instead of relying on signals. 

So with this, Spark Connect streaming and ODBC can coexist in the same process without issues. Pretty important for services that need both traditional ODBC connectivity (faster than gRPC) and Spark Connect's DataFrame capabilities (easier to use than ODBC).

caldempsey commented:
I looked into the panic `signal 13 received but handler not on signal stack... fatal error: non-Go code set up signal handler without SA_ONSTACK flag` that was happening in my application w/ ODBC and Spark Connect loaded side by side. Turns out it only occurs when running the Databricks ODBC driver alongside Spark Connect in the same process - run either one alone and everything works fine.

I added some socket monitoring with `lsof` into the Go app (get its own PID and pass it through and print stats on a ticker) to see what was going on and found both libraries maintain their own connections to Databricks through port 443:

```bash
ODBC Connections:    172.17.0.2:42562->44.234.192.45:443
                     172.17.0.2:42568->44.234.192.45:443
Spark Connect:       172.17.0.2:60452->44.234.192.46:443
                     172.17.0.2:41742->44.234.192.45:443
```

I _**THINK**_ the issue is becaauuuuse POSIX signals get delivered to the entire process, not individual sockets. I'm guessing the ODBC driver (`/opt/simba/spark/lib/64/libsparkodbc_sb64.so`) installs signal handlers when it loads, but doesn't use the `SA_ONSTACK` flag that Go requires. So when either library hits a broken pipe, the kernel sends SIGPIPE to the process, the ODBC handler catches it, and Go panics because the handler isn't set up correctly for Go's stack management.

With Spark Connect v40's new streaming and context cancellation support, SIGPIPE happens more often - cancel a context mid-stream and you'll likely trigger it. The ODBC driver's process-wide signal handler intercepts these signals regardless of which library's socket actually broke, while Spark Connect seems to handle the termination gracefully. The 'fix' here is really sweeping the issue under the rug by adding `signal.Ignore(syscall.SIGPIPE)` at startup e.g.

```go
package main 

func init() {
	signal.Ignore(syscall.SIGPIPE)
}
```

When you call signal.Ignore(syscall.SIGPIPE), the entire process ignores SIGPIPE signals. The kernel simply doesn't deliver them to the Go process at all. Both libraries already handle EPIPE errors properly through return values, so they don't need the signal. A C++ dev I know (who I can't mention) said this is actually the modern way to handle broken pipes - check return values instead of relying on signals. 

So with this, Spark Connect streaming and ODBC can coexist in the same process without issues. Pretty important for services that need both traditional ODBC connectivity (faster than gRPC) and Spark Connect's DataFrame capabilities (easier to use than ODBC).

grundprinzip reviewed (commented):
While working on a side project, I think this change could come in handy. I think the code is in a good spot, but would be good to add an integration test as well to show as well how to use the API.

caldempsey commented:
@grundprinzip Thanks Martin, yes I will get to the feedback as soon as I can, this Saturday is my next best available op. I've been vacationing + had a surgery, sorry for the delay!

caldempsey commented:
I’ve made the changes we discussed @grundprinzip @zeroshade. Few thoughts:

- I haven’t run a full end-to-end test yet to confirm that we’re truly fetching data at client pace without leaks. The channel-based implementation already handled this case well by blocking explicitly, so I’ve passed the sequences wrapping the business logic in the same way. My main concern is converting the push iterator in `ToRecordIterator` into a `Pull2` upstream. Hopefully this... just works, and won’t introduce leaks, but I’m still new to iterators in Go (we’ve been without them for 15 years!), so we better test E2E.

- One thing I don’t love is [this detail from the docs](https://pkg.go.dev/iter#Pull2): _“It is an error to call next or stop from multiple goroutines simultaneously.”_. That limitation feels a bit restrictive, but it does simplify the model, and I don't see any reason you need more than one go-routine doing the consumption here anyway.

- Iterators bring some nice benefits: being synchronous, guaranteeing no data race if the iterator and function body share state (although arguably modifying a data-structure while you are iterating it is slightly cursed and hard to read), has a fraction of the cost of channels + spinning up a goroutine, supports loop breaking control flow (return inside a loop body and break), while with channels you need to pass-in a cancellation channel (or context same thing) and manually close it when you break out, and here you can just... for loop over the Sequence type gracefully. That's pretty neat! 

- But again, worried that this could be a bit of a leaky abstraction as we cross the context boundaries of gRPC,  client.go, and the Pull2 type (which I've aliased to make sure we know its a 'Pull' sequence, as they both share the same type, frustratingly), as I've never used Sequences before today.

So please don't crucify me if this isn't exactly right!!!!!

caldempsey commented:
I’ve made the changes we discussed @grundprinzip @zeroshade. Few thoughts:

- I haven’t run a full end-to-end test yet to confirm that we’re truly fetching data at client pace without leaks. The channel-based implementation already handled this case well by blocking explicitly, so I’ve passed the sequences wrapping the business logic in the same way. My main concern is converting the push iterator in `ToRecordIterator` into a `Pull2` upstream. Hopefully this... just works, and won’t introduce leaks, but I’m still new to iterators in Go (we’ve been without them for 15 years!), so we better test E2E.

- One thing I don’t love is [this detail from the docs](https://pkg.go.dev/iter#Pull2): _“It is an error to call next or stop from multiple goroutines simultaneously.”_. That limitation feels a bit restrictive, but it does simplify the model, and I don't see any reason you need more than one go-routine doing the consumption here anyway.

- Iterators bring some nice benefits: being synchronous, guaranteeing no data race if the iterator and function body share state (although arguably modifying a data-structure while you are iterating it is slightly cursed and hard to read), has a fraction of the cost of channels + spinning up a goroutine, supports loop breaking control flow (return inside a loop body and break), while with channels you need to pass-in a cancellation channel (or context same thing) and manually close it when you break out, and here you can just... for loop over the Sequence type gracefully. That's pretty neat! 

- But again, worried that this could be a bit of a leaky abstraction as we cross the context boundaries of gRPC,  client.go, and the Pull2 type (which I've aliased to make sure we know its a 'Pull' sequence, as they both share the same type, frustratingly), as I've never used Sequences before today.

- I settled on the name 'Stream' with the Pull2 type: 'Stream' to convey the intent 'We want to stream records from Spark over time' and the type to convey 'We want the client to determine the rate Rows get consumed'. But maybe ToLocalIterator with the `Seq2` type (not a `Pull2`) would make more sense,  I'm just not exactly sure if that's the semantics we want to go for (again unfamiliar with the exact implications). Hence noted in `dataframe.go`:

```
	// StreamRows exposes a pull-based iterator over Arrow record batches from Spark types.RowPull2.
	// No rows are fetched from Spark over gRPC until the previous one has been consumed.
	// It provides no internal buffering: each Row is produced only when the caller
	// requests it, ensuring client back-pressure is respected.
	// types.RowPull2 is single use (can only be ranged once).
	StreamRows(ctx context.Context) (types.RowPull2, error)
```

So please don't crucify me if this isn't exactly right!!!!!

caldempsey commented:
> While working on a side project, I think this change could come in handy. I think the code is in a good spot, but would be good to add an integration test as well to show as well how to use the API.

Will take a look at writing the integration test next step, missed this comment!

caldempsey commented on spark/client/client.go:
```diff
@@ -434,6 +443,151 @@ func (c *ExecutePlanClient) ToTable() (*types.StructType, arrow.Table, error) {
+		}()
+
+		// Explicitly needed when tracking re-attachable execution.
+		c.done = false
```
I'm not exactly sure how this is _supposed_ to be implemented, any similar code or resources you can point out before I run over this again?

zeroshade replied:
wouldn't this be a race condition? Should we be locking around accessing `c.done`?

caldempsey replied:
Taking another look at this. I've spliced in the approach from ToTable(), originally written by @grundprinzip. Instead of writing to the shared c.done field, ToRecordSequence now tracks completion with a local done variable inside the closure.

After EOF, it checks if c.opts.ReattachExecution && !done and yields an error, the same way ToTable() does. This removes the race condition because nothing shared is being mutated, and it behaves correctly in both reattachable and non-reattachable modes.

I do think we should DRY this up eventually, but judiciously. I've kept the code WET for now intentionally. I want both code paths to remain directly comparable until I fully understand the domain. Having @grundprinzip's original logic in ToTable() sitting side by side against my equivalent in ToRecordSequence makes it much easier to reason about correctness and spot differences.

So rather than DRY it up in this PR, I'd like to take ownership of a fast follow-up where I consolidate ToRecordSequence and ToTable once this gets merged.

That said, happy to take a crack. I'm being cautious since this is an open source project, and I'd rather not introduce a negative diff that risks breaking users when the current approach is correct, comparable, and maintains parity between the two separate critical paths.

caldempsey commented on spark/sql/types/rowiterator.go:
```diff
@@ -0,0 +1,88 @@
+			// Treat io.EOF as clean termination (don’t forward).
+			if errors.Is(err, io.EOF) {
+				return
+			}
```
Since we're also pruning downstream maybe remove this

caldempsey commented:
@grundprinzip @zeroshade Added a Spark test that works fine locally, but looks like the Github action is failing on `wget -q https://dlcdn.apache.org/spark/spark-4.0.0/spark-4.0.0-bin-hadoop3.tgz`.

I'd normally address concerns in separate PRs, but I've bumped the version inline to keep things moving. Also added an integration test demonstrating how this works with unary RPCs. All current feedback should be addressed. Let me know if anything else needs attention 🙂 

Separately, I'd love to connect more with the community on this project. If there are any channels where folks discuss features, use cases, or development, I'd like to be part of them. I've been productionizing Spark Connect Go for a while now (a vote of confidence), including my own implementation of this feature with Databricks to synchronize the control loop between Spark requests and infrastructure (previously asynchronous Spark submit jobs). Would be good to share and listen to insights and experiences in the wild and smooth problems out! 

Thanks ✌️

zeroshade commented on spark/client/client.go:
```diff
@@ -368,6 +369,15 @@ func (c *ExecutePlanClient) ToTable() (*types.StructType, arrow.Table, error) {
+			if _, ok := resp.ResponseType.(*proto.ExecutePlanResponse_ResultComplete_); ok {
+				fmt.Println("DEBUG: Got ResultComplete!")
+			}
+		}
```
Probably shouldn't keep these debug lines in here. At minimum, use `log.Println` if we're not gonna switch to something like zerolog yet.

caldempsey replied:
ah, a blunder, thanks

caldempsey replied:
removed the debug lines

zeroshade commented on spark/client/client.go:
```diff
@@ -434,6 +444,110 @@ func (c *ExecutePlanClient) ToTable() (*types.StructType, arrow.Table, error) {
+// ToRecordSequence returns a single Seq2 iterator
+func (c *ExecutePlanClient) ToRecordSequence(ctx context.Context) iter.Seq2[arrow.Record, error] {
+	// Return Seq2 iterator that directly yields results as they arrive, upstream callers can convert this as needed
+	iterator := func(yield func(arrow.Record, error) bool) {
```
just do `return func(yield func(arrow.Record, error) bool) {` instead of `iterator := ....` and `return iterator`

zeroshade commented on spark/sql/types/rowiterator.go:
```diff
@@ -0,0 +1,88 @@
+				if !yield(row, nil) {
+					return
+				}
+			}
```
It feels like we could simplify this if `ReadArrowRecordToRows` performed the release inside the iterator etc.

Something like:

```go
for row := range rowIterFromRecord(rec) {
    if !yield(row, nil) {
        return
    }
}
```

caldempsey replied:
So i've extracted ReadArrowRecordToRows, and that let us range over it directly as you suggested.

There was an error check that needs to be handled since `ReadArrowRecordToRows` can still fail, so made that explicit:

```go
for row, err := range rowIterFromRecord(rec) {
    if err != nil {
         _ = yield(nil, err)
        return
    }
    if !yield(row, nil) {
        return
    }
}
```

zeroshade commented on spark/sql/dataframe.go:
```diff
@@ -936,6 +942,17 @@ func (df *dataFrameImpl) ToArrow(ctx context.Context) (*arrow.Table, error) {
+
+	seq2 := responseClient.ToRecordSequence(ctx)
+
+	return types.NewRowPull2(ctx, seq2), nil
```
Why not just use `iter.Pull2`?

caldempsey replied:
So, as I mentioned briefly, this is my first time using Go iterators. I was originally inspired  to build a `Pull` abstraction from the [Go range functions blog post](https://go.dev/blog/range-functions#pull-iterators) to give callers on-demand 'pull' semantics over the tuple. I used that as my 'north star', so to speak. 

But yeah, I think the `iter.Pull2` conversion here is overengineering: looking at this a few months down the line this is just creating a push→pull→push, and so a roundtrip that seems completely redundant and less concise. Seems the underlying gRPC stream is inherently single-use, so we don't need to guard termination either.

I'll deal with this by taking the useful part, the EOF handler and bubble it directly into `NewRowSequence` and drop the `iter.Pull2` conversion.

caldempsey replied:
Done, let me know if this looks like the right idiom to you. Apologies for the greenness.

zeroshade reviewed (commented):
Most of my comments are nitpicks, there's a few things that I think are pretty important but overall this looks pretty good to me!

caldempsey commented:
Updated the PR addressing feedback. Removed `NewRowPull2` entirely and consolidated it into `NewRowSequence`. Fixed the c.done race condition by having `ToRecordSequence` splice against `ToTable`. Kept things WET intentionally for now to avoid a negative diff against ToTable, will DRY up in a follow-up. Other comments addressed. Marking as ready for re-review.

zeroshade commented on spark/sql/types/rowiterator.go:
```diff
@@ -0,0 +1,67 @@
+				return
+			}
+			if rec == nil {
+				_ = yield(nil, errors.New("expected arrow.Record to contain non-nil Rows, got nil"))
```
The error here isn't quite correct, should probably be "expected non-nil arrow.Record" or something to that effect. It's not a case of "non-nil Rows" but rather the entire arrow.Record object is nil

caldempsey replied:
eagle eyes, good catch.

zeroshade commented on spark/sql/types/rowiterator.go:
```diff
@@ -0,0 +1,67 @@
+				}
+				if !yield(row, nil) {
+					return
+				}
```
Maybe something like this?

```suggestion
                cont := yield(row, err)
                if err != nil || !cont {
					return
				}
```

caldempsey replied:
yeah!  i wasn't sure if we wanted the explicit form of this where you make the `err != nil` flow _painfully_ clear 

```go
for row, err := range rowIterFromRecord(rec) {
    if !yield(row, err) || err != nil {
        return
    }
}
```

is another way to hash this. but honestly? i like your version most because it preserves the semantics from a human readability standpoint better.

```go
for row, err := range rowIterFromRecord(rec) {
    cont := yield(row, err)
    if err != nil || !cont {
        return
    }
}
```

will update

zeroshade reviewed (approved):
Just a couple comments, but otherwise looks great to me!
